File: go/iam/ListUsers/ListUsers.go
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: MIT-0
// snippet-start:[iam.go.list_users]
package main

// snippet-start:[iam.go.list_users.imports]
import (
    "flag"
    "fmt"

    "github.com/aws/aws-sdk-go/aws/session"
    "github.com/aws/aws-sdk-go/service/iam"
)
// snippet-end:[iam.go.list_users.imports]

// GetUsers retrieves a list of your IAM users
// Inputs:
//     sess is the current session, which provides configuration for the SDK's service clients
//     maxItems is the maximum number of users to return
// Output:
//     If success, the list of IAM users and nil
//     Otherwise, nil an error from the call to ListUsers
func GetUsers(sess *session.Session, maxItems *int64) (*iam.ListUsersOutput, error) {
    // snippet-start:[iam.go.list_users.call]
    svc := iam.New(sess)

    result, err := svc.ListUsers(&iam.ListUsersInput{
        MaxItems: maxItems,
    })
    // snippet-end:[iam.go.list_users.call]
    if err != nil {
        return nil, err
    }

    return result, nil
}

func main() {
    // snippet-start:[iam.go.list_users.args]
    maxUsers := flag.Int64("m", 10, "The maximum number of users to return")
    flag.Parse()

    if *maxUsers < int64(0) {
        *maxUsers = int64(10)
    }
    // snippet-end:[iam.go.list_users.args]

    // snippet-start:[iam.go.list_users.session]
    sess := session.Must(session.NewSessionWithOptions(session.Options{
        SharedConfigState: session.SharedConfigEnable,
    }))
    // snippet-end:[iam.go.list_users.session]

    result, err := GetUsers(sess, maxUsers)
    if err != nil {
        fmt.Println("Got an error retrieving users:")
        fmt.Println(err)
        return
    }

    for _, user := range result.Users {
        fmt.Println(*user.UserName+" created on", *user.CreateDate)
    }
}
// snippet-end:[iam.go.list_users]

File: swift/example_code/iam/ListUsers/Sources/ListUsers.swift
//
// Swift Example: ListUsers
//
// An example showing how to use the Amazon Identity and Access Management (IAM)
// `IAMClient` function `listUsers()`.
//
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0.

// snippet-start:[iam.swift.listusers.example]
// snippet-start:[iam.swift.listusers.main.imports]
import Foundation
import ArgumentParser
import AWSIAM
// snippet-end:[iam.swift.listusers.main.imports]

/// The command line arguments and options available for this
/// example command.
// snippet-start:[iam.swift.listusers.command]
struct ExampleCommand: ParsableCommand {
    static var configuration = CommandConfiguration(
        commandName: "listusers",
        abstract: "Lists all the users on your AWS account.",
        discussion: """
        """
    )

    // snippet-start:[iam.swift.listusers.command.getusernames]
    /// Return an array of strings listing the users available on the active
    /// Amazon IAM account.
    ///
    /// - Returns: An array of strings, each giving the name of one IAM user.
    func getUserNames(session: UserSession) async throws -> [String] {
        /// An array to store the found user names.
        var userNames: [String] = []

        /// A marker string used by IAM to indicate where the next batch of
        /// users begins. `nil` indicates the first user.
        var marker: String? = nil

        /// Whether or not the current batch of users is incomplete.
        var isTruncated: Bool = false

        // Call `listUsers()` repeatedly until the returned list
        // is not truncated.
        repeat {
            let input = ListUsersInput(marker: marker)
            let output = try await session.listUsers(input: input)

            guard let users = output.users else {
                continue
            }

            // Get the returned users' names and add them to the list of
            // users in `userNames`.
            for user in users {
                guard let name = user.userName else {
                    continue
                }

                userNames.append(name)

                // Set up for the next batch, if there is one.
                marker = output.marker
                isTruncated = output.isTruncated
            }
        } while isTruncated == true

        return userNames
    }
    // snippet-end:[iam.swift.listusers.command.getusernames]

    /// Called by ``main()`` to do the actual running of the AWS
    /// example.
    // snippet-start:[iam.swift.listusers.command.runasync]
    func runAsync() async throws {
        let session = try IAMUserSession()
        let userNames = try await getUserNames(session: session)

        print("Found \(userNames.count) users")
        for user in userNames {
            print("  \(user)")
        }
    }
    // snippet-end:[iam.swift.listusers.command.runasync]
}
// snippet-end:[iam.swift.listusers.command]

//
// Main program entry point.
//
// snippet-start:[iam.swift.listusers.main]
@main
struct Main {
    static func main() async {
        let args = Array(CommandLine.arguments.dropFirst())

        do {
            let command = try ExampleCommand.parse(args)

            try await command.runAsync()
        } catch {
            ExampleCommand.exit(withError: error)
        }
    }    
}
// snippet-end:[iam.swift.listusers.main]
// snippet-end:[iam.swift.listusers.example]